spec/services/form_validator_spec.rb

```diff
@@ -0,0 +1,198 @@
+require 'rails_helper'
+
+describe FormValidator do
+
+  let(:form) { create :form }
+  let(:element) { create :form_element, form: form, label: 'test' }
+
+  before :each do
+    @params = {form_id: element.form_id}
+  end
+
+  describe "errors and save" do
+
+    describe "adds an error if" do
+
+      after :each do
+        expect(@validator.valid?).to eq false
+      end
+      
+      describe "required field" do
+
+        before :each do
+          element.update_attributes({required: true})
+        end
+
+        after :each do
+          @validator = FormValidator.new(@params)
+          expect(@validator.errors).to eq ({test: ['is required']})
+        end
+
+        it "does not appear" do
+          @params.merge!({})
+        end
+
+        it "is nil" do
+          @params.merge!(test: nil)
+        end
+
+        it "is false" do
+          @params.merge!(test: false)
+        end
+
+        it "is empty string" do
+          @params.merge!(test: "")
+        end
+      end
+
+      describe "email field" do
+
+        before :each do
+          element.update_attributes({data_type: "email"})
+        end
+
+        after :each do
+          @validator = FormValidator.new(@params)
+          expect(@validator.errors).to eq ({test: [I18n.t("validation.is_invalid_email")]})
+        end
+
+        it "is a basic sentence" do
+          @params.merge!(test: "I'm not an email!")
+        end
+
+        it "is missing the TLD" do
+          @params.merge!(test: "neal@sumofus")
+        end
+
+        it "has two @ symbols" do
+          @params.merge!(test: "this@that@other.com")
+        end
+      end
+
+      describe "phone" do
+
+        before :each do
+          element.update_attributes({data_type: "phone"})
+        end
+
+        after :each do
+          @validator = FormValidator.new(@params)
+          expect(@validator.errors).to eq ({test: [I18n.t("validation.is_invalid_phone")]})
```

Comment:
```
irb(main):002:0> I18n.t("blerg")
=> "translation missing: en.blerg"
```
I think that's fine then. I'm testing that the method pulls in the translated text, not that there is translated text. The test will pass or fail the same whether or not the translation is there, and I think that's the desired behavior.